pkg/spanconfig/spanconfigsqlwatcher/sql_watcher.go

```diff
@@ -0,0 +1,453 @@
+
+	handleEvent := func(ctx context.Context, ev *roachpb.RangeFeedValue) {
+		var descID descpb.ID
+		{
+			// Decode the descriptor ID from the key.
+			tbl := systemschema.ZonesTable
+			types := []*types.T{tbl.PublicColumns()[0].GetType()}
+			startKeyRow := make([]rowenc.EncDatum, 1)
+			_, matches, _, err := rowenc.DecodeIndexKey(
+				s.codec, tbl, tbl.GetPrimaryIndex(),
+				types, startKeyRow, nil, ev.Key,
+			)
+			if err != nil || !matches {
+				logcrash.ReportOrPanic(
+					ctx,
+					&s.settings.SV,
+					"failed to decode key in system.zones: %v",
+					ev.Key,
+				)
+				return
+			}
+			alloc := &rowenc.DatumAlloc{}
+			if err := startKeyRow[0].EnsureDecoded(types[0], alloc); err != nil {
+				logcrash.ReportOrPanic(
+					ctx,
+					&s.settings.SV,
+					"failed to decode key in system.zones: %v",
+					ev.Key,
+				)
+				return
+			}
+			descID = descpb.ID(tree.MustBeDInt(startKeyRow[0].Datum))
+		}
+
+		select {
+		case <-ctx.Done():
+		case updatesCh <- descID:
+		}
+	}
+	rf, err := s.rangeFeedFactory.RangeFeed(
+		ctx,
+		"sql-watcher-zones-rangefeed",
+		descriptorTableSpan,
+		s.clock.Now(),
+		handleEvent,
+	)
+	if err != nil {
+		return nil, err
+	}
+	s.stopper.AddCloser(rf)
+
+	log.Infof(ctx, "established range feed over system.zones table")
+	return updatesCh, nil
+}
+
+// getAllAffectedDescriptors returns a list of IDs that need to have their span
+// configurations refreshed because the given id changed.
+func getAllAffectedIDs(
+	ctx context.Context, id descpb.ID, txn *kv.Txn, descsCol *descs.Collection,
+) (descpb.IDs, error) {
+	desc, err := descsCol.GetImmutableDescriptorByID(ctx, txn, id, tree.CommonLookupFlags{
+		IncludeDropped: true,
+	})
+	if err != nil {
+		if errors.Is(err, catalog.ErrDescriptorNotFound) {
+			return descpb.IDs{id}, nil
+		}
+		return nil, err
+	}
+	switch desc.DescriptorType() {
+	case catalog.Table:
+		return descpb.IDs{id}, nil
+	case catalog.Database:
+	default:
+		return nil, errors.AssertionFailedf("expected either database or table, but found descriptor of type [%s]", desc.DescriptorType())
+	}
+	// There's nothing to do here if the database has been dropped. If the
+	// database was non-empty those objects will have their own rangefeed events.
+	if desc.Dropped() {
+		return nil, nil
+	}
+	descriptors, err := descsCol.GetAllDescriptorsInDatabase(ctx, txn, id)
+	if err != nil {
+		return nil, err
+	}
+	ret := make(descpb.IDs, 0, len(descriptors))
+	for _, desc := range descriptors {
+		ret = append(ret, desc.GetID())
+	}
+	return ret, nil
+}
+
+// onDescIDUpdate generates the span configurations for the given descriptor ID
+// and sends them to the provided updatesCh.
+func (s *SQLWatcher) onDescIDUpdate(
+	ctx context.Context, descID descpb.ID, updatesCh chan spanconfig.Update,
+) error {
+	var updates []spanconfig.Update
+	if err := descs.Txn(
+		ctx,
+		s.settings,
+		s.leaseManager,
+		s.ie,
+		s.DB,
+		func(ctx context.Context, txn *kv.Txn, descsCol *descs.Collection) error {
```

Comment:
Isn't this closure retryable? We cannot append to updates blindly -- reset on each invocation.